README.md
# H-99: Ninety-Nine Haskell Problems

[Ninety-Nine Haskell Problems](https://wiki.haskell.org/H-99:_Ninety-Nine_Haskell_Problems)

## The problems

These problems have been split into 11 parts, for ease of access.

- Questions 1 to 10: Lists
- Questions 11 to 20: Lists, continued
- Questions 21 to 28: Lists again
- Questions 31 to 41: Arithmetic
- Questions 46 to 50: Logic and codes
- Questions 54A to 60: Binary trees
- Questions 61 to 69: Binary trees, continued
- Questions 70B to 73: Multiway trees
- Questions 80 to 89: Graphs
- Questions 90 to 94: Miscellaneous problems
- Questions 95 to 99: Miscellaneous problems, continued

Though the problems number from 1 to 99, there are some gaps and some additions marked with letters. There are actually only 88 problems.


ChangeLog.md
# Changelog for 99-Haskell-Problems

## Unreleased changes


src/Part01.hs
{-
 - Part 01
 - Lists
 -}
import Data.List (group)


{-
 - Problem 1: Find the last element of a list.
 -}
myLast :: [a] -> a
myLast [] = error "empty list"
myLast [x] = x
myLast (_:xs) = myLast xs

-- -- test cases:
-- main = do
--     putStrLn "Problem 1"
--     print $ myLast [1, 2, 3, 4] == 4
--     print $ myLast ['x','y','z'] == 'z'


{-
 - Problem 2: Find the last but one element of a list.
 -}
myButLast :: [a] -> a
myButLast [x] = error "only one element in list"
myButLast [x, _] = x
myButLast (_:xs) = myButLast xs

myButLast' = last . init

-- -- test cases:
-- main = do
--     putStrLn "Problem 2"
--     print $ myButLast [1,2,3,4] == 3
--     print $ myButLast ['a'..'z'] == 'y'


{-
 - Problem 3: Find the K'th element of a list. The fst element in the list is number 1.
 -}
elementAt :: [a] -> Int -> a
elementAt xs num = xs !! (num - 1)

-- -- test cases:
-- main = do
--     putStrLn "Problem 3"
--     print $ elementAt [1,2,3] 2 == 2
--     print $ elementAt "haskell" 5 == 'e'


{-
 - Problem 4: Find the number of elements of a list.
 -}
myLength :: [a] -> Int
myLength xs = sum [1 | _ <- xs]

myLength' :: [a] -> Int
myLength' = sum . map (\_ -> 1)
-- myLength' xs = sum . map (\_ -> 1) xs

myLength'' :: [a] -> Int
myLength'' [] =  0
myLength'' (_:xs) =  1 + myLength'' xs

-- -- test cases:
-- main = do
--     putStrLn "Problem 4"
--     print $ myLength [123, 456, 789] == 3
--     print $ myLength "Hello, world!" == 13


{-
 - Problem 5: Reverse a list.
 -}
myReverse :: [a] -> [a]
myReverse [] = []
myReverse (x:xs) = myReverse xs ++ [x]  -- naive way LOL

myReverse' :: [a] -> [a]
myReverse' = foldl (flip (:)) []
-- myReverse' = foldl (\acc x -> x : acc) []

-- -- test cases:
-- main = do
--     putStrLn "Problem 5"
--     print $ myReverse' "A man, a plan, a canal, panama!" == "!amanap ,lanac a ,nalp a ,nam A"
--     print $ myReverse' [1,2,3,4] == [4,3,2,1]


{-
 - Problem 6: Find out whether a list is a palindrome.
 - A palindrome can be read forward or backward; e.g. (x a m a x).
 -}
isPalindrome :: Eq a => [a] -> Bool
isPalindrome [] = True
isPalindrome [_] = True
isPalindrome xs = if (head xs == last xs) then
                      isPalindrome $ (tail . init) xs
                  else False
-- isPalindrome xs = (head xs) == (last xs) && (isPalindrome' $ init $ tail xs)

isPalindrome'' xs = xs == (reverse xs)  -- more efficient or not? memeory/intermediate var/speed

-- Here's one using foldr and zipWith.
palindrome''' :: Eq a => [a] -> Bool
palindrome''' xs = foldr (&&) True $ zipWith (==) xs (reverse xs)
palindrome'''' xs = and $ zipWith (==) xs (reverse xs) -- same, but easier

-- -- test cases:
-- main = do
--     putStrLn "Problem 6"
--     print $ isPalindrome [1,2,3] == False
--     print $ isPalindrome "madamimadam" == True
--     print $ isPalindrome [1,2,4,8,16,8,4,2,1] == True


{-
 - Problem 7 (**) Flatten a nested list structure.
 - Transform a list, possibly holding lists as elements into a `flat' list by replacing each list with its elements (recursively).
 - We have to define a new data type, because lists in Haskell are homogeneous.
 -}
data NestedList a = Elem a | List [NestedList a] deriving Show

flatten :: NestedList a -> [a]
flatten (Elem x) = [x]
flatten (List x) = concatMap flatten x

-- test cases:
main = do
    putStrLn "Problem 7"
    print $ flatten (Elem 5) == [5]
    print $ flatten (List [Elem 1, List [Elem 2, List [Elem 3, Elem 4], Elem 5]]) == [1, 2, 3, 4, 5]
    -- print $ flatten (List []) == []  -- FIXME: error?


{-
 - Problem 8 (**) Eliminate consecutive duplicates of list elements.
 - If a list contains repeated elements they should be replaced with a single copy of the element.
 - The order of the elements should not be changed.
 -}
compress :: Eq a => [a] -> [a]
compress xs = foldr (\x acc -> if x == head acc then acc else x:acc) [last xs] (init xs)
-- this one is not so efficient, because it pushes the whole input onto the "stack" before doing anything else.
-- why?

-- Note that GHC erases the Maybes, producing efficient code.
compress' :: Eq a => [a] -> [a]
compress' xs = foldr skipDups (const []) xs Nothing
    where skipDups x acc a@(Just q) | (x == q) = acc a
          skipDups x acc _ = x : acc (Just x)

compress'' :: Eq a => [a] -> [a]
compress'' = map head . group

-- -- test cases:
-- main = do
--     putStrLn "Problem 8"
--     print $ compress' "aaaabccaadeeee" == "abcade"


{-
 - Problem 9 (**) Pack consecutive duplicates of list elements into sublists.
 - If a list contains repeated elements they should be placed in separate sublists.
 -}
pack :: Eq a => [a] -> [[a]]
pack [] = []
pack xs = let (first, rest) = span (== head xs) xs in first : pack rest

pack' :: Eq a => [a] -> [[a]]
pack' [] = []
pack' xs = takeWhile (== head xs) xs : pack (dropWhile (== head xs) xs)

-- -- test cases:
-- main = do
--     putStrLn "Problem 9"
--     print $ pack' ['a', 'a', 'a', 'a', 'b', 'c', 'c', 'a', 'a', 'd', 'e', 'e', 'e', 'e'] == ["aaaa", "b", "cc", "aa", "d", "eeee"]


{-
 - Problem 10 (*) Run-length encoding of a list.
 - Use the result of problem P09 to implement the so-called run-length encoding data compression method.
 - Consecutive duplicates of elements are encoded as lists (N E) where N is the number of duplicates of the element E.
 -}
encode :: Eq a => [a] -> [(Int, a)]
encode xs = map (\x -> (length x, head x)) $ pack xs  -- pack = group

-- -- test cases:
-- main = do
--     putStrLn "Problem 10"
--     print $ encode "aaaabccaadeeee" == [(4, 'a'), (1, 'b'), (2, 'c'), (2, 'a'),(1, 'd'), (4, 'e')]


src/Part02.hs
{-
 - Part 02
 - Lists, continued
 -}
import Data.List (group)
import Control.Applicative ((<**>))

{-
-- Problem 11 (*) Modified run-length encoding.
-- Modify the result of problem 10 in such a way that if an element has no duplicates it is simply copied into the result list.
 - Only elements with duplicates are transferred as (N E) lists.
 -}
data ListItem a = Single a | Multiple Int a deriving (Show, Eq)

encodeModified :: Eq a => [a] -> [ListItem a]
encodeModified xs = map countDups $ group xs
    where countDups xs = let dups_length = length xs
                             element = head xs in
                         if dups_length == 1
                         then Single element
                         else Multiple dups_length element
-- too long?

-- shorter
encodeModified' :: Eq a => [a] -> [ListItem a]
encodeModified' xs = map (encodeHelper . (\dups -> (length dups, head dups))) $ group xs
    where encodeHelper (1, x) = Single x
          encodeHelper (n, x) = Multiple n x

-- much shorter
encodeModified'' :: Eq a => [a] -> [ListItem a]
encodeModified'' xs = [y | x <- group xs,
                       let y = if (length x) == 1 then Single (head x) else Multiple (length x) (head x)]

-- -- test cases:
-- main = do
--     putStrLn "Problem 11"
--     print $ encodeModified' "aaaabccaadeeee" == [Multiple 4 'a', Single 'b', Multiple 2 'c', Multiple 2 'a', Single 'd', Multiple 4 'e']


{-
 - Problem 12 (**) Decode a run-length encoded list.
 - Given a run-length code list generated as specified in problem 11.
 - Construct its uncompressed version.
-}
decodeModified :: [ListItem a] -> [a]
decodeModified = concatMap decodeHelper
    where decodeHelper (Single x) = [x]
          decodeHelper (Multiple c x) = replicate c x

-- use uncurry. nice solution :)
decodeModified' :: [ListItem a] -> [a]
decodeModified' = concatMap (uncurry replicate . toTuple)
    where toTuple :: ListItem a -> (Int, a)
          toTuple (Single x) = (1, x)
          toTuple (Multiple n x) = (n, x)

-- a naïve solution with foldl
decodeModified'' :: [ListItem a] -> [a]
decodeModified'' = foldl (\acc e -> case e of Single x -> acc ++ [x]; Multiple n x -> acc ++ replicate n x) []

-- -- test cases:
-- main = do
--     putStrLn "Problem 12"
--     print $ decodeModified' [Multiple 4 'a', Single 'b', Multiple 2 'c',
--         Multiple 2 'a', Single 'd', Multiple 4 'e'] == "aaaabccaadeeee"


{-
 - Problem 13 (**) Run-length encoding of a list (direct solution).
 - Implement the so-called run-length encoding data compression method directly.
 - I.e. don't explicitly create the sublists containing the duplicates, as in problem 9, but only count them.
 - As in problem P11, simplify the result list by replacing the singleton lists (1 X) by X.
 -}
-- decodeDirect :: Eq a => [a] -> [ListItem a]
-- decodeDirect = ???

-- -- test cases:
-- main = do
--     putStrLn "Problem 13"
--     print $ encodeDirect "aaaabccaadeeee" == [Multiple 4 'a', Single 'b', Multiple 2 'c', Multiple 2 'a', Single 'd', Multiple 4 'e']


{-
 - Problem 14 (*) Duplicate the elements of a list.
 -}
dupli :: [a] -> [a]
dupli = concatMap (\x -> [x, x])

-- using the list instance of `Applicative`
dupli' = (<**> [id, id])

-- -- test cases:
-- main = do
--     putStrLn "Problem 14"
--     print $ dupli' [1, 2, 3] == [1, 1, 2, 2, 3, 3]


{-
 - Problem 15 (**) Replicate the elements of a list a given number of times.
 -}
repli :: [a] -> Int -> [a]
repli xs n = concatMap (replicate n) xs

-- or, in Pointfree style:
repli' :: [a] -> Int -> [a]
repli' = flip $ concatMap . replicate

-- or, using the list monad:
repli'' :: [a] -> Int -> [a]
repli'' xs n = xs >>= replicate n

-- -- test cases:
-- main = do
--     putStrLn "Problem 15"
--     print $ repli "abc" 3 == "aaabbbccc"


{-
 - Problem 16 (**) Drop every N'th element from a list.
 -}
dropEvery :: [a] -> Int -> [a]
dropEvery xs n = foldr (\x acc -> snd x : acc) [] $ filter (uncurry notDrop) $ zip [0..(length xs)-1] xs
    where notDrop idx _ = (idx+1) `mod` n /= 0
-- improve: dropEvery xs n = map snd . filter (uncurry notDrop) $ zip [0..(length xs)-1] xs

-- shortest of my idea
dropEvery' :: [a] -> Int -> [a]
dropEvery' xs n = map snd $ filter (\(i, _) -> i `mod` n /= 0) $ zip [1..] xs

-- Using zip:
dropEvery'' xs n = map fst $ filter ((n/=) . snd) $ zip xs (cycle [1..n])

-- Using comprehensions
dropEvery''' :: [a] -> Int -> [a]
dropEvery''' xs n = [i | (i, c) <- (zip xs [1,2..]), (mod c n) /= 0]

-- Using unzip
dropEvery'''' :: [a] -> Int -> [a]
dropEvery'''' xs k = (snd . unzip . filter (\(i, _) -> i `mod` k /= 0) . zip [1..]) xs
-- Type signature is slightly different from the problem, but equivalent and shorter.
-- dropEvery :: Int -> [a] -> [a]
-- dropEvery k = snd . unzip . filter (\(i, _) -> i `mod` k /= 0) . zip [1..]

-- -- test cases:
-- main = do
--     putStrLn "Problem 16"
--     print $ dropEvery'''' "abcdefghik" 3
--     print $ dropEvery'''' "abcdefghik" 3 == "abdeghk"


{-
 - Problem 17 (*) Split a list into two parts; the length of the first part is given.
 - Do not use any predefined predicates.
 -}
-- split :: [a] -> Int -> ([a], [a])
-- split xs loc = (take loc xs, drop loc xs)
-- Oh, no. These should clearly be considered "predefined predicates".
split :: [a] -> Int -> ([a], [a])
split [] _ = ([], [])
split all@(x:xs) n
    | n <= 0 = ([], all)
    | n == 1 = ([x], xs)  -- actually you could remove this condition but one more recursion
    | otherwise = (x:first, rest)
    where (first, rest) = split xs (n - 1)

-- Another foldl solution without defining tuple extractors:
split'' :: [a] -> Int -> ([a], [a])
split'' lst n = snd $ foldl helper (0, ([], [])) lst
    where helper (i, (left, right)) x = if i >= n
                                        then (i + 1, (left, right ++ [x]))
                                        else (i + 1, (left ++ [x], right))

-- -- test cases:
-- main = do
--     putStrLn "Problem 17"
--     print $ split "abcdefghik" 3
--     print $ split "abcdefghik" 3 == ("abc", "defghik")


{-
 - Problem 18 (**) Extract a slice from a list.
 - Given two indices, i and k, the slice is the list containing the elements between
 - the i'th and k'th element of the original list (both limits included).
 - Start counting the elements with 1.
 -}
slice :: [a] -> Int -> Int -> [a]
slice xs i k | i > 0 = take (k - i + 1) $ drop (i - 1) xs
-- notice `i>0` guard condition

-- A solution using list comprehension:
slice :: [a] -> Int -> Int -> [a]
slice xs i k = [x | (x, j) <- zip xs [1..k], i <= j]

-- Zip, filter, unzip:
slice'' :: [a] -> Int -> Int -> [a]
slice'' xs i k = fst $ unzip $ filter ((>= i) . snd) $ zip xs [1..k]

-- test cases:
main = do
    putStrLn "Problem 18"
    print $ slice ['a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'k'] 3 7 == "cdefg"


{-
 - Problem 19 (**) Rotate a list N places to the left.
 - Hint: Use the predefined functions length and (++).
 -}
rotate :: [a] -> Int -> [a]
rotate [] _ = []
rotate xs 0 = xs
rotate xs shift = reverse $ reverse first ++ reverse second
    where (first, second) = if shift > 0 then splitAt shift xs else splitAt (length xs + shift) xs

-- using cycle
rotate' :: [a] -> Int -> [a]
rotate' xs n = take len . drop (n `mod` len) . cycle $ xs
    where len = length xs

-- without mod
rotate'' :: [a] -> Int -> [a]
rotate'' xs n = take (length xs) $ drop (length xs + n) $ cycle xs

-- -- test cases:
-- main = do
--     putStrLn "Problem 19"
--     print $ rotate ['a','b','c','d','e','f','g','h'] 3 == "defghabc"
--     print $ rotate ['a','b','c','d','e','f','g','h'] (-2) == "ghabcdef"


{-
 -  Problem 20 (*) Remove the K'th element from a list.
 - (Note that not only returns the residue list, but also returns the deleted element.)
 -}
removeAt :: Int -> [a] -> (a, [a])
removeAt _ [] = error "empty list"
removeAt k xs = (xs !! (k-1), take (k-1) xs ++ drop k xs)

-- a safe version of that
removeAt' :: Int -> [a] -> (Maybe a, [a])
removeAt' n xs | n > 0 && n <= length xs = (Just (xs !! idx), take idx xs ++ drop n xs)
               | otherwise = (Nothing, xs)
               where idx = n - 1

-- another solution that also uses Maybe to indicate failure:
removeAt'' :: Int -> [a] -> (Maybe a, [a])
removeAt'' _ [] = (Nothing, [])
removeAt'' 0 xs = (Nothing, xs)
removeAt'' nr xs | nr > length xs = (Nothing, xs)
                 | otherwise = (Just (xs !! nr), fst splitted ++ (tail . snd) splitted)
                where splitted = splitAt nr xs

-- similar, point-free style:
removeAt''' :: Int -> [a] -> (a, [a])
removeAt''' n = (\(a, b) -> (head b, a ++ tail b)) . splitAt (n - 1)

-- -- test cases:
-- main = do
--     putStrLn "Problem 20"
--     print $ removeAt 2 "abcd" == ('b', "acd")


src/Part03.hs
{-
 - Part 03
 - Lists again
 -}

{-
 - Problem 21 Insert an element at a given position into a list.
 -}
insertAt :: a -> [a] -> Int -> [a]
insertAt x xs idx = first ++ x : second
    where (first, second) = splitAt (idx - 1) xs

-- -- test cases:
-- main = do
--     putStrLn "Problem 21"
--     print $ insertAt 'X' "abcd" 2 == "aXbcd"


{-
 - Problem 22 Create a list containing all integers within a given range.
 -}
range :: Int -> Int -> Int -> [Int]
range start stop 1 = [start..stop]  -- Note that range list in haskell doesn't follow [,) rule.
range start stop step = [start, start + step .. real_stop]  -- Python-like range
    where real_stop = if (stop-start) `mod` step == 0 then stop - step else stop
-- TODO: how to implement/define optional arguments?

range' :: Int -> Int -> Int -> [Int]
range' start stop 1 = take (stop - start + 1)  $ iterate (+1) start
range' start stop step = take ((stop - start + step) `div` step) $ iterate (+ step) start

-- pred/succ and handle start > stop condition
range'' :: (Ord a, Enum a) => a -> a -> [a]
range'' a b | (a == b) = [a]
range'' a b = a:range'' ((if a < b then succ else pred) a) b

-- use scanl
range''' :: Int -> Int -> [Int]
range''' l r = scanl (+) l (replicate (l - r) 1)

-- Since there's already syntactic sugar for ranges, there's usually no reason to define a function like 'range' in Haskell.
-- In fact, the syntactic sugar is implemented using the `enumFromTo` function, which is exactly what 'range' should be.

-- -- test cases:
-- main = do
--     putStrLn "Problem 22"
--     -- print $ range 4 9 == [4, 5, 6, 7, 8, 9]
--     print $ range 4 9 1 == [4, 5, 6, 7, 8, 9]
--     print $ range 0 10 2 == [0, 2, 4, 6, 8]
--     print $ range 0 9 2 == [0, 2, 4, 6, 8]

{-
 - Problem 23 Extract a given number of randomly selected elements from a list.
 - import System.Random (rnd_select)
 -}



-- test cases:
main = do
    putStrLn "Problem 23"
    print rnd_select "abcdefgh" 3 >>= putStrLn == "eda"



{-
 - Problem 24 Lotto: Draw N different random numbers from the set 1..M.
 - import System.Random (diff_select)
 -}

-- -- test cases:
-- main = do
--     putStrLn "Problem 24"
--     print $ diff_select 6 49 == [23, 1, 17, 33, 21, 37]


{-
 - Problem 25 Generate a random permutation of the elements of a list.
 - import System.Random (rnd_permu)
 -}

-- -- test cases:
-- main = do
--     putStrLn "Problem 25"
--     print $ rnd_permu "abcdef" == "badcef"


{-
 - Problem 26 (**) Generate the combinations of K distinct objects chosen from the N elements of a list
 - In how many ways can a committee of 3 be chosen from a group of 12 people? We all know that
    there are C(12,3) = 220 possibilities (C(N,K) denotes the well-known binomial coefficients).
 - For pure mathematicians, this result may be great.
 - But we want to really generate all the possibilities in a list.
 -}
combinations :: Int -> [a] -> [[a]]
combinations k xs |


-- -- test cases:
-- main = do
--     putStrLn "Problem 26"
--     print $ combinations 3 "abcdef" == ["abc", "abd", "abe", ...]



{-
 - Problem 27 Group the elements of a set into disjoint subsets.
 - a) In how many ways can a group of 9 people work in 3 disjoint subgroups of 2, 3 and 4 persons?
 - Write a function that generates all the possibilities and returns them in a list.
 - b) Generalize the above predicate in a way that we can specify a list of group sizes and the predicate will return a list of groups.
 - Note that we do not want permutations of the group members; i.e. ((ALDO BEAT) ...) is the same solution as ((BEAT ALDO) ...).
 - However, we make a difference between ((ALDO BEAT) (CARLA DAVID) ...) and ((CARLA DAVID) (ALDO BEAT) ...).

 - You may find more about this combinatorial problem in a good book on discrete mathematics under the term "multinomial coefficients".
 -}

-- -- test cases:
-- main = do
--     putStrLn "Problem 27"
--     print $ group [2,3,4] ["aldo", "beat", "carla", "david", "evi", "flip", "gary", "hugo", "ida"] == [[["aldo", "beat"], ["carla", "david", "evi"], ["flip", "gary", "hugo", "ida"]], ...]
-- -- (altogether 1260)
--     print $ group [2,2,5] ["aldo", "beat", "carla", "david", "evi", "flip", "gary", "hugo", "ida"] == [[["aldo", "beat"], ["carla", "david"], ["evi", "flip", "gary", "hugo", "ida"]], ...]
-- -- (altogether 756)


{-
 - Problem 28 Sorting a list of lists according to length of sublists
 - a) We suppose that a list contains elements that are lists themselves.
 - The objective is to sort the elements of this list according to their length.
 - E.g. short lists first, longer lists later, or vice versa.
 -}



-- -- test cases:
-- main = do
--     putStrLn "Problem 28"

-- lsort ["abc", "de", "fgh", "de", "ijkl", "mn", "o"] == ["o", "de", "de", "mn", "abc", "fgh", "ijkl"]


{-
 - (b) Again, we suppose that a list contains elements that are lists themselves.
 - But this time the objective is to sort the elements of this list according to their length frequency;
 - i.e., in the default, where sorting is done ascendingly, lists with rare lengths are placed first, others with a more frequent length come later.
 -}

-- -- test cases:
-- main = do
--     putStrLn "Problem 21"

-- lfsort ["abc", "de", "fgh", "de", "ijkl", "mn", "o"] == ["ijkl", "o", "abc", "fgh", "de", "de", "mn"]


src/Part05.hs
{-
 - Part 05
 - Logic and Codes
 -}

{-
 - Problem 46 (**) Define predicates and/2, or/2, nand/2, nor/2, xor/2, impl/2 and equ/2 (for logical equivalence) which succeed or fail according to the result of their respective operations;
 - e.g. and(A,B) will succeed, if and only if both A and B succeed.
 - A logical expression in two variables can then be written as in the following example: and(or(A,B),nand(A,B)).
 - Now, write a predicate table/3 which prints the truth table of a given logical expression in two variables.
 -}


-- test cases:
main = do
    putStrLn "Problem 46"
    print $ table (\a b -> (and' a (or' a b)))
{-
    True  True  True
    True  False True
    False True  False
    False False False
 -}


{-
 - Problem 47 (*) Truth tables for logical expressions (2).
 - Continue problem P46 by defining and/2, or/2, etc as being operators.
 - This allows to write the logical expression in the more natural way, as in the example: A and (A or not B).
 - Define operator precedence as usual; i.e. as in Java.
 -}


-- test cases:
main = do
    putStrLn "Problem 47"
    print $ table2 (\a b -> a `and'` (a `or'` not b))

True  True  True
True  False True
False True  False
False False False


{-
 - Problem 48 (**) Truth tables for logical expressions (3).
 - Generalize problem P47 in such a way that the logical expression may contain any number of logical variables.
 - Define table/2 in a way that table(List, Expr) prints the truth table for the expression Expr, which contains the logical variables enumerated in List.
 -}

-- test cases:
main = do
    putStrLn "Problem 48"
    print $ tablen 3 (\[a,b,c] -> a `and'` (b `or'` c) `equ'` a `and'` b `or'` a `and'` c)
-- infixl 3 `equ'`
True  True  True  True
True  True  False True
True  False True  True
True  False False True
False True  True  True
False True  False True
False False True  True
False False False True

-- infixl 7 `equ'`
True  True  True  True
True  True  False True
True  False True  True
True  False False False
False True  True  False
False True  False False
False False True  False
False False False False


{-
 - Problem 49 (**) Gray codes.
 - An n-bit Gray code is a sequence of n-bit strings constructed according to certain rules. For example,

    n = 1: C(1) = ['0','1'].
    n = 2: C(2) = ['00','01','11','10'].
    n = 3: C(3) = ['000','001','011','010',´110´,´111´,´101´,´100´].

 - Find out the construction rules and write a predicate with the following specification:

    % gray(N,C) :- C is the N-bit Gray code

 - Can you apply the method of "result caching" in order to make the predicate more efficient, when it is to be used repeatedly?
 -}


-- test cases:
main = do
    putStrLn "Problem 4"
    print $ gray 3 == ["000", "001", "011", "010", "110", "111", "101", "100"]


{-
 - Problem 50 (***) Huffman codes.
 - We suppose a set of symbols with their frequencies, given as a list of fr(S,F) terms.
 - Example: [fr(a,45), fr(b,13), fr(c,12), fr(d,16), fr(e,9), fr(f,5)].
 - Our objective is to construct a list hc(S,C) terms, where C is the Huffman code word for the symbol S.
 - In our example, the result could be Hs = [hc(a,'0'), hc(b,'101'), hc(c,'100'), hc(d,'111'), hc(e,'1101'), hc(f,'1100')] [hc(a,'01'),...etc.].
 - The task shall be performed by the predicate huffman/2 defined as follows:

    % huffman(Fs,Hs) :- Hs is the Huffman code table for the frequency table Fs

 -}


-- test cases:
main = do
    putStrLn "Problem 4"
    print $ huffman [('a', 45), ('b', 13), ('c', 12), ('d', 16), ('e', 9), ('f', 5)] == [('a', "0"), ('b', "101"), ('c', "100"), ('d', "111"), ('e', "1101"), ('f', "1100")]


src/Part04.hs
{-
 - Part 04
 - Arithmetic
 -}


{-
 - Problem 31 (**) Determine whether a given integer number is prime.
 -}


-- test cases:
main = do
    putStrLn "Problem 31"
    print $ isPrime 7 == True


{-
 - Problem 32 (**) Determine the greatest common divisor of two positive integer numbers.
 - Use Euclid's algorithm.
 -}

-- test cases:
main = do
    putStrLn "Problem 32"
    print $ [myGCD 36 63, myGCD (-3) (-6), myGCD (-3) 6] == [9, 3, 3]

{-
 - Problem 33 (*) Determine whether two positive integer numbers are coprime.
 - Two numbers are coprime if their greatest common divisor equals 1.
 -}

-- test cases:
main = do
    putStrLn "Problem 33"
    print $ coprime 35 64 == True


{-
 - Problem 34 (**) Calculate Euler's totient function phi(m).
 - Euler's so-called totient function phi(m) is defined as the number of positive integers r (1 <= r < m) that are coprime to m.
 - Example: m = 10: r = 1, 3, 7, 9; thus phi(m) = 4.
 - Note the special case: phi(1) = 1.
 -}

-- test cases:
main = do
    putStrLn "Problem 34"
    print $ totient 10 == 4


{-
 - Problem 35 (**) Determine the prime factors of a given positive integer.
 - Construct a flat list containing the prime factors in ascending order.
 -}

-- test cases:
main = do
    putStrLn "Problem 35"
    print $ primeFactors 315 == [3, 3, 5, 7]


{-
 - Problem 36 (**) Determine the prime factors of a given positive integer.
 - Construct a list containing the prime factors and their multiplicity.
 -}

-- test cases:
main = do
    putStrLn "Problem 36"
    print $ prime_factors_mult 315 == [(3, 2), (5, 1), (7, 1)]


{-
 - Problem 37 (**) Calculate Euler's totient function phi(m) (improved).
 - See problem 34 for the definition of Euler's totient function.
 - If the list of the prime factors of a number m is known in the form of problem 36 then
    the function phi(m) can be efficiently calculated as follows: Let ((p1 m1) (p2 m2) (p3 m3) ...)
    be the list of prime factors (and their multiplicities) of a given number m.
 - Then phi(m) can be calculated with the following formula:

    phi(m) = (p1 - 1) * p1 ** (m1 - 1) *
             (p2 - 1) * p2 ** (m2 - 1) *
             (p3 - 1) * p3 ** (m3 - 1) * ...

 - Note that a ** b stands for the b'th power of a.
 -}


{-
 - Problem 38 (*) Compare the two methods of calculating Euler's totient function.
 - Use the of problems 34 and 37 to compare the algorithms.
 - Take the number of reductions as a measure for efficiency.
 - Try to calculate phi(10090) as an example.
 - (no solution required)
 -}



{-
 - Problem 39 (*) A list of prime numbers.
 - Given a range of integers by its lower and upper limit,
    construct a list of all prime numbers in that range.
 -}


-- test cases:
main = do
    putStrLn "Problem 39"
    print $ primesR 10 20 == [11,13,17,19]


{-
 - Problem 40 (**) Goldbach's conjecture.
 - Goldbach's conjecture says that every positive even number greater than 2 is the sum of two prime numbers.
 - Example: 28 = 5 + 23.
 - It is one of the most famous facts in number theory that has not been proved to be correct in the general case.
 - It has been numerically confirmed up to very large numbers (much larger than we can go with our Prolog system).
 - Write a predicate to find the two prime numbers that sum up to a given even integer.
 -}

-- test cases:
main = do
    putStrLn "Problem 40"
    print $ goldbach 28 == (5, 23)


{-
 - Problem 41 (**) Given a range of integers by its lower and upper limit,
    print a list of all even numbers and their Goldbach composition.
 - In most cases, if an even number is written as the sum of two prime numbers, one of them is very small.
 - Very rarely, the primes are both bigger than say 50.
 - Try to find out how many such cases there are in the range 2..3000.
 -}


-- test cases:
main = do
    putStrLn "Problem 41"
    print $ goldbachList 9 20 == [(3, 7), (5, 7), (3, 11), (3, 13), (5, 13), (3, 17)]
    print $ goldbachList' 4 2000 50 == [(73, 919), (61, 1321), (67, 1789), (61, 1867)]
